gcloud-java-contrib/gcloud-java-nio/src/main/java/com/google/gcloud/storage/contrib/nio/CloudStorageFileSystemProvider.java

```diff
@@ -79,6 +82,34 @@
 @AutoService(FileSystemProvider.class)
 public final class CloudStorageFileSystemProvider extends FileSystemProvider {
 
+  private static class LazyPathIterator extends AbstractIterator<Path> {
+    private Iterator<Blob> blobIterator;
+    private Filter<? super Path> filter;
+    private CloudStorageFileSystem fileSystem;
+
+    LazyPathIterator(CloudStorageFileSystem fileSystem, Iterator<Blob> blobIterator, Filter<? super Path> filter) {
+      this.blobIterator = blobIterator;
+      this.filter = filter;
+      this.fileSystem = fileSystem;
+    }
+
+    @Override
+    protected Path computeNext() {
+      while (blobIterator.hasNext()) {
+        Path path = fileSystem.getPath(blobIterator.next().name());
+        try {
+          if (filter.accept(path)) {
+            return path;
+          }
+        } catch (IOException ex) {
+          throw new DirectoryIteratorException(ex);
+        }
+      }
+      return endOfData();
+    }
+  }
+
+
```

Comment:
That's funny, I only see one line in the IDE. Well, I removed it, we'll see how that goes.